File: backend/internal/model/organization.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// Organization represents an organization in the system
type Organization struct {
	ID               uuid.UUID `json:"id" db:"id"`
	ClerkOrgID       string    `json:"clerkOrgId" db:"clerk_org_id"`
	Name             string    `json:"name" db:"name"`
	Slug             string    `json:"slug" db:"slug"`
	LogoURL          *string   `json:"logoUrl,omitempty" db:"logo_url"`
	SubscriptionTier string    `json:"subscriptionTier" db:"subscription_tier"`
	MaxMembers       int       `json:"maxMembers" db:"max_members"`
	MaxProjects      int       `json:"maxProjects" db:"max_projects"`
	IsActive         bool      `json:"isActive" db:"is_active"`
	CreatedAt        time.Time `json:"createdAt" db:"created_at"`
	UpdatedAt        time.Time `json:"updatedAt" db:"updated_at"`
}

// CreateOrganizationParams represents the parameters for creating a new organization
type CreateOrganizationParams struct {
	ClerkOrgID       string
	Name             string
	Slug             string
	LogoURL          *string
	SubscriptionTier string
	MaxMembers       int
	MaxProjects      int
}

// UpdateOrganizationParams represents the parameters for updating an organization
type UpdateOrganizationParams struct {
	Name             *string
	Slug             *string
	LogoURL          *string
	SubscriptionTier *string
	MaxMembers       *int
	MaxProjects      *int
	IsActive         *bool
}

File: packages/zod/src/organization.ts
import { z } from "./z.js";

export const OrganizationSchema = z.object({
    id: z.string().uuid(),
    clerk_org_id: z.string(),
    name: z.string(),
    slug: z.string(),
    logo_url: z.string().url().nullable(),
    subscription_tier: z.enum(["free", "pro", "enterprise"]),
    max_members: z.number(),
    max_projects: z.number(),
    is_active: z.boolean(),
    created_at: z.string(),
    updated_at: z.string(),
}).openapi("Organization");

export const UpdateOrganizationSchema = z.object({
    name: z.string().optional(),
    slug: z.string().optional(),
    logo_url: z.string().url().optional(),
    subscription_tier: z.enum(["free", "pro", "enterprise"]).optional(),
    max_members: z.number().optional(),
    max_projects: z.number().optional(),
    is_active: z.boolean().optional(),
});

export const CreateOrganizationSchema = z.object({
    name: z.string().min(1),
    slug: z.string().min(1),
    logo_url: z.string().url().optional(),
});

export const AddMemberSchema = z.object({
    user_id: z.string().uuid(),
    role: z.string(),
});

export const UpdateMemberRoleSchema = z.object({
    role: z.string(),
});

export type Organization = z.infer<typeof OrganizationSchema>;
export type UpdateOrganization = z.infer<typeof UpdateOrganizationSchema>;
export type CreateOrganization = z.infer<typeof CreateOrganizationSchema>;

export const OrganizationMemberSchema = z.object({
    id: z.string().uuid(),
    organization_id: z.string().uuid(),
    user_id: z.string().uuid(),
    role: z.string(),
    joined_at: z.string(),
}).openapi("OrganizationMember");

export const OrganizationMemberWithDetailsSchema = OrganizationMemberSchema.extend({
    user_email: z.string().email().optional(),
    user_name: z.string().optional(),
    organization_name: z.string().optional(),
}).openapi("OrganizationMemberWithDetails");

export type OrganizationMember = z.infer<typeof OrganizationMemberSchema>;
export type OrganizationMemberWithDetails = z.infer<typeof OrganizationMemberWithDetailsSchema>;
export type AddMember = z.infer<typeof AddMemberSchema>;
export type UpdateMemberRole = z.infer<typeof UpdateMemberRoleSchema>;
